File: src/main/java/cn/afeibaili/mchat/MChat.java
package cn.afeibaili.mchat;

import cn.afeibaili.mchat.config.Config;
import cn.afeibaili.mchat.socket.message.MessageManager;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.text.Style;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.TextFormatting;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.common.config.Configuration;
import net.minecraftforge.event.ServerChatEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
import net.minecraftforge.fml.common.event.FMLServerStartingEvent;
import net.minecraftforge.fml.common.event.FMLServerStoppingEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.PlayerEvent;

import java.io.File;

import static cn.afeibaili.mchat.MChat.mchatSystem;

@Mod(
        modid = MChat.MODID,
        name = MChat.NAME,
        version = MChat.VERSION,
        acceptableRemoteVersions = "*"
)
public class MChat {
    public static final String MODID = "mchat";
    public static final String NAME = "MChat";
    public static final String VERSION = "1.0";

    public static MChatSystem mchatSystem = new MChatSystem();

    @EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        MChatModConfig.load(event.getSuggestedConfigurationFile());
        MinecraftForge.EVENT_BUS.register(this);
        MinecraftForge.EVENT_BUS.register(new Listener());
    }

    @EventHandler
    public void onCommonSetup(FMLServerStartingEvent event) {
        MessageManager<TextFormatting, MinecraftServer> messageManager = new MessageManager<TextFormatting, MinecraftServer>() {
            @Override
            public TextFormatting getFormat() {
                return TextFormatting.GRAY;
            }

            @Override
            public void sendToMc(String message, TextFormatting format) {
                event.getServer().getPlayerList().getPlayers().forEach(player -> {
                    player.sendMessage(new TextComponentString(message).setStyle(new Style().setColor(format)));
                });
            }
        };

        mchatSystem.buildManager(messageManager);
        mchatSystem.connect(ConfigLoader.config);
    }

    @EventHandler
    public void onCommonStop(FMLServerStoppingEvent event) {
        MChatSystem.stop();
    }
}


//class Listener implements MessageListener<ServerChatEvent, PlayerEvent.PlayerLoggedInEvent, PlayerEvent.PlayerLoggedOutEvent> {
class Listener {
    String name = ConfigLoader.config.getName();

    @SubscribeEvent
    public void onMessage(ServerChatEvent event) {
        mchatSystem.getMessageManager()
                .sendToGroup("[" + name + "] " + event.getPlayer().getName() + ": " + event.getMessage());
    }

    @SubscribeEvent
    public void onPlayerIn(PlayerEvent.PlayerLoggedInEvent event) {
        mchatSystem.getMessageManager()
                .sendLoginInOut(event.player.getName() + "加入了" + name + "服务器");
    }

    @SubscribeEvent
    public void onPlayerOut(PlayerEvent.PlayerLoggedOutEvent event) {
        mchatSystem.getMessageManager()
                .sendLoginInOut(event.player.getName() + "退出了" + name + "服务器");
    }
}

class ConfigLoader {
    public static final Config config = new Config(
            MChatModConfig.token,
            MChatModConfig.address,
            MChatModConfig.port,
            MChatModConfig.name
    );
}

class MChatModConfig {
    public static String address;
    public static int port;
    public static String token;
    public static String name;

    public static void load(File file) {
        Configuration configuration = new Configuration(file);
        configuration.load();

        address = configuration.getString("address", "general", "192.168.1.8", "server address");
        port = configuration.getInt("port", "general", 33393, 1, 65535, "server port");
        token = configuration.getString("token", "general", "c5c5d6ce-", "server token");
        name = configuration.getString("name", "general", "未命名", "server name");

        if (configuration.hasChanged()) {
            configuration.save();
        }
    }
}

File: forge-1.20.1/src/main/kotlin/online/afeibaili/mchat/MChat.kt
package online.afeibaili.mchat

import com.mojang.blaze3d.platform.NativeImage
import com.mojang.brigadier.arguments.StringArgumentType
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraft.commands.Commands
import net.minecraft.network.chat.ClickEvent
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraftforge.client.event.RegisterClientCommandsEvent
import net.minecraftforge.common.ForgeConfigSpec
import net.minecraftforge.event.ServerChatEvent
import net.minecraftforge.event.entity.player.PlayerEvent
import net.minecraftforge.event.server.ServerStartingEvent
import net.minecraftforge.event.server.ServerStoppingEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.config.ModConfig
import online.afeibaili.mchat.MChat.Companion.mchatSystem
import online.afeibaili.mchat.config.Config
import online.afeibaili.mchat.listener.MessageListener
import online.afeibaili.mchat.socket.message.MessageManager
import thedarkcolour.kotlinforforge.forge.FORGE_BUS
import thedarkcolour.kotlinforforge.forge.registerConfig
import java.io.Closeable
import java.io.InputStream
import java.net.URL
import java.util.*

@Mod(MChat.ID)
class MChat {
    companion object {
        const val ID = "mchat"
        val mchatSystem: MChatSystem = MChatSystem()
    }

    init {
        FORGE_BUS.register(this)
        FORGE_BUS.register(Listener())
        registerConfig(ModConfig.Type.COMMON, MChatModConfig.spec)
    }

    @SubscribeEvent
    fun onCommonSetup(event: ServerStartingEvent) {
        val manager: MessageManager<ChatFormatting, MinecraftServer> =
            object : MessageManager<ChatFormatting, MinecraftServer>() {
                override var formatting: ChatFormatting = ChatFormatting.GRAY
                override fun sendToMC(message: String, formatting: ChatFormatting) {
                    event.server.playerList.players.forEach { player ->
                        player.sendSystemMessage(Component.literal(message).withStyle(formatting))
                    }
                }

                override fun sendImageToMC(message: String, formatting: ChatFormatting) {
                    event.server.playerList.players.forEach { player ->
                        player.sendSystemMessage(
                            Component.literal("[点击查看图片消息]").withStyle { style ->
                                style.withColor(ChatFormatting.BLUE)
                                    .withClickEvent(ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, "/image $message"))
                            }
                        )
                    }
                }
            }


        mchatSystem.buildManager(manager)
        mchatSystem.connect(ConfigLoader.config)
    }

    @SubscribeEvent
    fun onCommonStop(event: ServerStoppingEvent) {
        MChatSystem.close()
    }

    @SubscribeEvent
    fun registerClientCommand(event: RegisterClientCommandsEvent) {
        event.dispatcher.register(
            Commands.literal("image")
                .then(
                    Commands.argument("url", StringArgumentType.greedyString()).executes { ctx ->
                        val image = ScreenImage(StringArgumentType.getString(ctx, "url"))
                        Minecraft.getInstance().setScreen(image)
                        return@executes 1
                    })
                .then(
                    Commands.literal("release").executes { ctx ->
                        ScreenImage.close()
                        Minecraft.getInstance().player?.sendSystemMessage(
                            Component.literal("已释放图片资源")
                                .withStyle(ChatFormatting.BLUE)
                        )
                        return@executes 1
                    }
                )
        )
    }
}

class Listener : MessageListener<ServerChatEvent, PlayerEvent.PlayerLoggedInEvent, PlayerEvent.PlayerLoggedOutEvent> {
    @SubscribeEvent
    override fun onMessage(event: ServerChatEvent) {
        mchatSystem.getMessageManager()
            .sendToGroup("[${ConfigLoader.config.name}] " + event.player.name.string + ": " + event.message.string)
    }

    @SubscribeEvent
    override fun onPlayerIn(event: PlayerEvent.PlayerLoggedInEvent) {
        mchatSystem.getMessageManager()
            .sendLoginInOut("${event.entity.name.string}加入了${ConfigLoader.config.name}服务器")
    }

    @SubscribeEvent
    override fun onPlayerOut(event: PlayerEvent.PlayerLoggedOutEvent) {
        mchatSystem.getMessageManager()
            .sendLoginInOut("${event.entity.name.string}退出了${ConfigLoader.config.name}服务器")
    }
}

object ConfigLoader {
    val config = Config(
        MChatModConfig.token.get(),
        MChatModConfig.address.get(),
        MChatModConfig.port.get(),
        MChatModConfig.name.get()
    )
}

object MChatModConfig {
    val builder = ForgeConfigSpec.Builder()

    val address: ForgeConfigSpec.ConfigValue<String> =
        builder.comment("mchat server url address")
            .define("address", "mc.afeibaili.cn")

    val port: ForgeConfigSpec.ConfigValue<Int> =
        builder.comment("mchat server port")
            .define("port", 33393)

    val token: ForgeConfigSpec.ConfigValue<String> =
        builder.comment("mchat server token")
            .define("token", "c5c5d6ce-")

    val name: ForgeConfigSpec.ConfigValue<String> =
        builder.comment("mchat server name")
            .define("name", "未命名")

    val spec: ForgeConfigSpec = builder.build()
}

class ScreenImage(url: String) : Screen(Component.literal("图片查看")), Closeable {
    val image: Image = getImageByUrl(url)

    override fun render(
        graphics: GuiGraphics,
        mouseX: Int,
        mouseY: Int,
        partialTick: Float,
    ) {

        runCatching {
            image.writeGraphics(graphics)
        }.onFailure { exception ->
            exception.printStackTrace()
            Minecraft.getInstance().player?.sendSystemMessage(
                Component.literal("图片损坏（按ESC关闭消息）：${exception.message}")
                    .withStyle(ChatFormatting.RED)
            )
        }

        super.render(graphics, mouseX, mouseY, partialTick)
    }

    override fun close() {
        image.close()
    }

    companion object {
        val map = HashMap<String, Image>()

        fun getImageByUrl(url: String): Image {
            val stream: InputStream = URL(url).openConnection().apply {
                connectTimeout = 5000
                readTimeout = 5000
            }.getInputStream()
            val nativeImage: NativeImage = NativeImage.read(stream)
            val texture = DynamicTexture(nativeImage)
            val uuid: String = UUID.randomUUID().toString()
            val location: ResourceLocation =
                Minecraft.getInstance().textureManager.register(uuid, texture)
            return Image(location, texture, nativeImage.width, nativeImage.height).apply {
                map[uuid] = this
            }
        }

        fun close() {
            map.forEach { _, image -> image.close() }
            map.clear()
        }
    }
}

class Image(val location: ResourceLocation, val texture: DynamicTexture, val width: Int, val height: Int) : Closeable {
    fun writeGraphics(graphics: GuiGraphics) {
        val mc = Minecraft.getInstance()

        val screenWidth = mc.window.guiScaledWidth
        val screenHeight = mc.window.guiScaledHeight

        val imageWidth = width.toFloat()
        val imageHeight = height.toFloat()

        val scale = minOf(
            screenWidth / imageWidth,
            screenHeight / imageHeight
        )

        val drawWidth = (imageWidth * scale).toInt()
        val drawHeight = (imageHeight * scale).toInt()

        val drawX = (screenWidth - drawWidth) / 2
        val drawY = (screenHeight - drawHeight) / 2

        graphics.blit(
            location,
            drawX,
            drawY,
            drawWidth,
            drawHeight,
            0f, 0f,
            width,
            height,
            width,
            height
        )
    }

    override fun close() {
        texture.close()
        Minecraft.getInstance().textureManager.release(location)
    }
}